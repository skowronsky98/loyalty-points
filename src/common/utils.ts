import crypto from 'crypto';

export const setDefault = <T>(
  value: T,
  defaultValue: Exclude<T, undefined>,
): Exclude<T, undefined> =>
  value === undefined ? defaultValue : (value as Exclude<T, undefined>);

export const hashEvent = (input: { eventName: string, sequence: number }): string => {
  const { eventName, sequence } = input;
  return crypto
    .createHash('sha256')
    .update(`${eventName}-${sequence}`)
    .digest('hex');
};
