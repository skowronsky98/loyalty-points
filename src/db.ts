import { EntityManager, MikroORM, Options } from "@mikro-orm/postgresql";
import mikroOrmConfig from "./mikro-orm.config.js";
import { LoyaltyPointsEntity } from "./entitites/loyalty-points.entity.js";
import { LoyaltyPointsRepository } from "./repos/loyalty-points.repo.js";
import { LoyaltyPointsDeptRepository } from "./repos/loyalty-points-dept.repo.js";
import { LoyaltyPointsDeptEntity } from "./entitites/loyalty-points-dept.entity.js";

export interface Services {
  orm: MikroORM;
  em: EntityManager;
  loyaltyPointsRepo: LoyaltyPointsRepository;
  loyaltyPointsDeptRepo: LoyaltyPointsDeptRepository;
}

let cache: Services;

export async function initORM(options?: Options): Promise<Services> {
  if (cache) {
    return cache;
  }

  const orm = await MikroORM.init({
    ...mikroOrmConfig,
    ...options,
  });

  return cache = {
    orm,
    em: orm.em,
    loyaltyPointsRepo: orm.em.getRepository(LoyaltyPointsEntity),
    loyaltyPointsDeptRepo: orm.em.getRepository(LoyaltyPointsDeptEntity),
  };
}
