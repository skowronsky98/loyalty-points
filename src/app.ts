import Fastify from 'fastify'
import { ENV } from './env/env.js'
import { initORM } from './db.js'
import { RequestContext } from '@mikro-orm/postgresql'
import { registerWebhookEvents } from './endpoints/webhook-events.endpoint.js'
import { registerPointsRoutes } from './endpoints/get-loyalty-points.endpoint.js'
import { registerConsumePoints } from './endpoints/consume-loyalty-points.endpoint.js'
import {
  jsonSchemaTransform,
  serializerCompiler,
  validatorCompiler,
  isResponseSerializationError,
  hasZodFastifySchemaValidationErrors,
  ZodTypeProvider,
} from 'fastify-type-provider-zod';
import fastifySwagger from '@fastify/swagger'
import fastifySwaggerUI from '@fastify/swagger-ui';

export const createApp = async () => {

  const db = await initORM();

  const app = Fastify().withTypeProvider<ZodTypeProvider>();

  app.setValidatorCompiler(validatorCompiler);
  app.setSerializerCompiler(serializerCompiler);

  app.register(fastifySwagger, {
    openapi: {
      info: {
        title: 'Loyalty Points Service',
        description: 'Sample backend service',
        version: '1.0.0',
      },
      servers: [],
    },
    transform: jsonSchemaTransform,
  });

  app.register(fastifySwaggerUI, {
    routePrefix: '/documentation',
  });


  app.addHook('onRequest', (request, reply, done) => {
    RequestContext.create(db.em, done);
  });

  app.setErrorHandler((error, request, reply) => {

    if (hasZodFastifySchemaValidationErrors(error)) {
      return reply.status(400).send({ error: 'Bad Request', statusCode: 400 });
    }

    if (isResponseSerializationError(error)) {
      return reply.code(500).send({
        error: 'Internal Server Error',
        statusCode: 500,
      })
    }

    app.log.error(error);

    return reply.code(500).send({
      error: 'Internal Server Error',
      statusCode: 500,
    })
  });

  app.addHook('onClose', async () => {
    await db.orm.close();
  });

  app.register(registerPointsRoutes);
  app.register(registerConsumePoints);
  app.register(registerWebhookEvents);

  await app.ready();
  await app.listen({ port: ENV.APP_PORT });

  console.log(`Documentation running at http://localhost:3000/documentation`);

  return app;
}

try {
  await createApp();
} catch (error) {
  console.error(error);
  process.exit(1);
}
