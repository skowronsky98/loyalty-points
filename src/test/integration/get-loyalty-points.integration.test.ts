import { FastifyInstance } from 'fastify'
import { afterAll, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import supertest from 'supertest'
import { initORM, Services } from '../../db.js'
import { createApp } from '../../app.js'
import { clearDatabase } from '../../test-utils/test.utils.js'
import { LoyaltyPointsEntity } from '../../entitites/loyalty-points.entity.js'
import { faker } from '@faker-js/faker';
import { addDays, subDays } from 'date-fns'
import { EntityManager } from '@mikro-orm/postgresql'
import { LoyaltyPointsDeptEntity } from '../../entitites/loyalty-points-dept.entity.js'

describe('get-loyalty-points', () => {
  let app: FastifyInstance
  let db: Services
  let em: EntityManager

  beforeAll(async () => {
    [app, db] = await Promise.all([createApp(), initORM()])
    em = db.orm.em.fork();
    await app.ready()
  })

  beforeEach(async () => {
    await clearDatabase(db.orm)
  });

  afterAll(async () => {
    app.close()
  })

  it('should not return expired loyalty points', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: faker.string.uuid(),
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: faker.string.uuid(),
      expiresAt: subDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    const response = await supertest(app.server)
      .get(`/${loyaltyPoints.customerId}/points`)
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    expect(response.body).toEqual({ pointsAvailable: 0 })
  })

  it('should not return loyalty points from returned orders', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: faker.string.uuid(),
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: faker.string.uuid(),
      expiresAt: subDays(new Date(), 1),
      isCancelled: true,
    });

    await em.persistAndFlush(loyaltyPoints);

    const response = await supertest(app.server)
      .get(`/${loyaltyPoints.customerId}/points`)
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    expect(response.body).toEqual({ pointsAvailable: 0 })
  })

  it('should return loyalty points', async () => {
    const loyaltyPoints = new LoyaltyPointsEntity({
      customerId: faker.string.uuid(),
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: faker.string.uuid(),
      expiresAt: addDays(new Date(), 1),
    });

    await em.persistAndFlush(loyaltyPoints);

    const response = await supertest(app.server)
      .get(`/${loyaltyPoints.customerId}/points`)
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    expect(response.body).toEqual({ pointsAvailable: 10 })
  })

  it('should return loyalty points sum', async () => {
    const customerId = faker.string.uuid();
    const loyaltyPoints = [new LoyaltyPointsEntity({
      customerId,
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: faker.string.uuid(),
      expiresAt: addDays(new Date(), 1),
    }), new LoyaltyPointsEntity({
      customerId,
      availablePoints: 10,
      pointsFromOrder: 10,
      orderId: faker.string.uuid(),
      expiresAt: addDays(new Date(), 1),
    })];

    await em.persistAndFlush(loyaltyPoints);

    const response = await supertest(app.server)
      .get(`/${customerId}/points`)
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    expect(response.body).toEqual({ pointsAvailable: 20 })
  })

  it('should return negative amount of loyalty points when customer is in dept', async () => {
    const customerId = faker.string.uuid();
    const loyaltyPointsDept = new LoyaltyPointsDeptEntity({
      customerId,
      points: 10,
    });

    await em.persistAndFlush(loyaltyPointsDept);

    const response = await supertest(app.server)
      .get(`/${customerId}/points`)
      .expect(200)
      .expect('Content-Type', 'application/json; charset=utf-8')

    expect(response.body).toEqual({ pointsAvailable: -10 })
  })
})
