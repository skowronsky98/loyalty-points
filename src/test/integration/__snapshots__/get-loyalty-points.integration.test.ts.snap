// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get-loyalty-points > snapshot 1`] = `
{
  "foo": "bar",
}
`;
